master/pkg/searcher/adaptive_asha.go

```diff
@@ -0,0 +1,94 @@
+package searcher
+
+import (
+	"math"
+	"sort"
+
+	"github.com/determined-ai/determined/master/pkg/model"
+)
+
+func getBracketMaxTrials(
+	maxTrials int, divisor float64, brackets []int) []int {
+	// This allocation will result in roughly equal total step budget
+	// allocated to each rung.
+	// Each bracket roughly requires numRungs * targetTrialSteps budget to evaluate
+	// divisor ^ (numRungs - 1) configurations.  Hence, we can compute the average
+	// budget per configuration for each bracket and back into the number of
+	// trials per bracket if we want roughly equal total step budget.
+	var bracketWeight []float64
+	totalWeight := 0.
+	for i, numRungs := range brackets {
+		bracketWeight = append(bracketWeight, math.Pow(divisor, float64(numRungs-1))/float64(numRungs))
+		totalWeight += bracketWeight[i]
+	}
+	var bracketTrials []int
+	allocated := 0
+	for i := 0; i < len(brackets); i++ {
+		bracketTrials = append(
+			bracketTrials, max(int(bracketWeight[i]/totalWeight*float64(maxTrials)), 1))
+
+		allocated += bracketTrials[i]
+	}
+	remainder := max(maxTrials-allocated, 0)
+	bracketTrials[0] += remainder
+	return bracketTrials
+}
+
+func getBracketMaxConcurrentTrials(
+	maxConcurrentTrials *int, divisor float64, maxTrials []int) []int {
+	// If maxConcurrentTrials is provided, we will split that evenly across brackets
+	// and fill remainder from most aggressive early stopping bracket to least.
+	// Otherwise, we will default to minimum of the maxTrials across brackets
+	// to guarantee roughly equal work between brackets.
+	var minTrials int
+	remainder := 0
+	numBrackets := len(maxTrials)
+	var bracketMaxConcurrentTrials []int
+	if maxConcurrentTrials == nil {
+		minTrials = max(maxTrials[numBrackets-1], int(divisor))
+	} else {
+		minTrials = *maxConcurrentTrials / numBrackets
+		remainder = *maxConcurrentTrials % numBrackets
+	}
+	for i := 0; i < numBrackets; i++ {
+		bracketMaxConcurrentTrials = append(bracketMaxConcurrentTrials, minTrials)
```

Comment:
Dividing maxConcurrentTrials evenly across brackets will make sure the brackets have approximately equal training resource so progress across brackets is roughly equal.  See example below:

4 slots for the experiment
queue: (bracket: 1, trial: 1, steps: 1), (bracket: 1, trial: 2, steps: 1), (bracket: 2, trial: 3, steps: 4), (bracket: 2, trial: 4, steps: 4)
events:
- time 1: job 1 and 2 finish and bracket 1 creates two new jobs (bracket: 1, trial: 5, steps: 1), (bracket: 1, trial: 6, steps: 1) that are added to queue. The two freed workers start working on these two jobs.
- time 2: job 5 and 6 finish and bracket 1 creates two new jobs (bracket: 1, trial: 7, steps: 1), (bracket: 1, trial: 8, steps: 1)
- time 3: job 7 and 8 finish and bracket 1 creates two new jobs (bracket: 1, trial: 9, steps: 1), (bracket: 1, trial: 10, steps: 1)
- time 4: job 3, 4, 9, 10 finish and bracket 1 and bracket 2 each create two new jobs.

Reply:
okay -- follow up is for my own curiosity

so is the expectation then that each bracket size is similar -- I would've assumed we wanted to have each bracket make proportional progress for it's size rather than equal progress

Reply:
the maxTrials for each bracket set by getBracketMaxTrials is such that the total training steps used by each bracket is about the same.  hence, making equal progress on the brackets would make it so that the brackets finish around the same time.